contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+    NetworkConfig network_config = 1;
+    UserNamespaceConfig user_namespace_config = 2;
+}
+
+message UpdateRuntimeConfigRequest {
+    RuntimeConfig runtime_config = 1;
+}
+
+message UpdateRuntimeConfigResponse {}
+
+message RuntimeConfigInfoRequest {}
+
+message RuntimeConfigInfoResponse {
+    RuntimeConfig runtime_config = 1
+}
+
+...
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+	// Network namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode network = 1;
+	// PID namespace for this container/sandbox.
+	// Note: The CRI default is POD, but the v1.PodSpec default is CONTAINER.
+	// The kubelet's runtime manager will set this to CONTAINER explicitly for v1 pods.
+	// Namespaces currently set by the kubelet: POD, CONTAINER, NODE
+	NamespaceMode pid = 2;
+	// IPC namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped IPC in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode ipc = 3;
+	// User namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped user namespace in the
+	// Kubernetes API. The container runtime has to run with user namespace enabled,
+	// and kubelet then forces the use of host user namespace.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceModeValue user = 4;
+}
+
+// NamespaceModeValue is the wrapper of NamespaceMode, to allow non-POD default.
+message NamespaceModeValue {
+	NamespaceMode value = 1;
+}
+```
+
+### Runtime Support
+- Docker: https://docs.docker.com/engine/security/userns-remap/
+    - Concerns:
+Docker API does not provide user-namespace mapping. Therefore to handle `RuntimeConfigInfo` API, changes will be done in `dockershim` to read system files, `/etc/subuid` and `/etc/subgid`, for figuring out default user-namespace mapping. It won’t be possible to use custom user/group name to `--userns-remap parameter`, the implementation will expect `dockremap`, which is the default username.
+- CRI-O: WIP https://github.com/nalind/cri-o/tree/idmapping
+
+### Implementation Roadmap
+#### Phase 1: Support in Kubelet, Alpha, [Target: Kubernetes v1.11]
+- Add feature gate `NodeUserNamespace`, disabled by default
+- Add new CRI API, `RuntimeConfigInfo()`
+- Add logic in Kubelet to handle pod creation which includes parsing RuntimeConfigInfo response and changing file-permissions in /var/lib/kubelet with learned userns mapping.
+- Add changes in dockershim to implement RuntimeConfigInfo() for docker runtime
+- Add changes in CRI-O to implement userns support and RuntimeConfigInfo() support
+- Unit test cases
+
+#### Phase 2: Beta Support [Target: Kubernetes v1.12]
+- PSP integration
+- To grow ExperimentalHostUserNamespaceDefaulting from experimental feature gate to a Kubelet flag
```

Comment:
should we wrap `ExperimentalHostUserNamespaceDefaulting` into `NodeUserNamespace` behavior?  It seems like you'd want them both.

Reply:
I'd much prefer for users to be explicit with `"hostUserNamespace": true` when needed, than rely on the implicit magic behaviour.

Reply:
I wonder about the experience there when the system knows that they need `hostUserNamespace: true`.  Seems like tying node level configuration knowledge to a user trying to run a pod.  

I consider it node level configuration because on a non-userns remap node not setting `hostUserNamespace` would work fine, on the userns remap node it would fail runtime validation during container creation.  The user expectation was just that they are allowed to user `hostNetwork` (or whatever) and it should work.

Reply:
> should we wrap ExperimentalHostUserNamespaceDefaulting into NodeUserNamespace behavior? It seems like you'd want them both.

only motivation for keeping them seperate is the flexibility that it provides to admin to choose whether he wants to allow privileged pods or not, in case NodeUserNamespace is enabled

Reply:
@pweil-, would you have some link to discussion which would explain why having one host namespace implies the user namespace to be host as well? Is that a general thing in the kernel which will always be true, or merely behaviour of the current Container Runtimes like docker?

Reply:
@adelton `"hostUserNamespace": true` seems complementary to the `ExperimentalHostUserNamespaceDefaulting`, rather an alternative to it. IMO, `ExperimentalHostUserNamespaceDefaulting` functionality is required even with `"hostUserNamespace": true` support.

Reply:
@pweil- I think it would make sense to fold them into one. I can't think of cases where users will be able to get away without enabling HostUserNamespaceDefaulting.

Reply:
> IMO, ExperimentalHostUserNamespaceDefaulting functionality is required even with "hostUserNamespace": true support.

In my testing, having the explicit `"hostUserNamespace": true` seems to work even with the unfixed `ExperimentalHostUserNamespaceDefaulting`. Granted, I've only checked if the process in the container is indeed running as uid 0 on the host, didn't do any functional testing.

Reply:
> Capabilities for a non-user namespace are checked against the owning user namespace so any time you're sharing a non-user namespace you need to share the owning user namespace in order to have permissions to use the non-user namespace

Thanks.

Reply:
for correctness, some others have noted that some usecases of non-host user ns and host pid/net/ipc DO work.  TIL https://github.com/kubernetes/community/pull/2067#discussion_r183835918

Reply:
@pweil- @adelton 

> Capabilities for a non-user namespace are checked against the owning user namespace so any time you're sharing a non-user namespace you need to share the owning user namespace in order to have permissions to use the non-user namespace

This is mostly true, though there are some exceptions (and there is no framework in the kernel to guarantee this, it's implemented for each namespace independently using `ns_capable`). But as a rule of thumb, yes this is pretty much true (except in the cases where you need a host capability or a capability in a parent namespace or something similar).